### DISCUSSION TITLE
Access HTTP only cookie in client component prerender phase

### DISCUSSION THREAD

**Discussion by ellemedit** [vercel/next.js#56395] [Help]
### Summary

Several APIs, utilized for fetching data referenced in Client Components, perform authentication via HTTP only cookies. However, when carrying out prerendering in a Node.js/Edge runtime from a Client Component, unlike in a browser, it's impossible to read the HTTP only cookies, thus making prerendering unfeasible.

Or there's alternative way to authorize API for client component? HTTP only cookie's kind of traditional and much safer than exposing JS runtime.

### Example

https://codesandbox.io/p/sandbox/lucid-zhukovsky-xvk382?file=%2Fapp%2Fpage.tsx%3A8%2C15

**Comment by icyJoseph** [comment]
During pre-rendering I guess you need to skip rendering of those parts that require some kind of Cookie to be present. Since during pre-rendering there's no transaction going on, there's no request headers to look into and get a cookie.

What you could do is to make the page dynamic, for example by using `export const dynamic = 'force-dynamic'`, that' stop pre-rendering of the page, and instead you'd render it on demand, as a request arrives to the server, and there you can look for cookies.

**Reply by ellemedit** [reply]
skipping prerendering is current solution as you said, but it makes unable to use streaming rendering in client component. I think there should be an escape hatch by framework.

**Reply by XiNiHa** [reply]
I guess the word _prerendering_ is the point of confusion here, @ellemedit is using the word to refer to SSR of client components, so there are transactions going on per each render.

**Reply by ellemedit** [reply]
what should I say for the term?

**Reply by XiNiHa** [reply]
Not sure 😂 AFAIK most people just call it SSR

**Reply by icyJoseph** [reply]
Aha, I think you might want to use these for:

- pre-rendering, for rendering done under `next build`
- on-demand rendering, for server side rendering done under `next start`
- CSR well, done in the browser

These are all also a bit nuanced, because I am mainly talking about React rendering, which happens on the first stage of the pixel pipeline. https://web.dev/articles/rendering-performance#the_pixel_pipeline

So what is it you want to do again?

**Reply by ellemedit** [reply]
No, Joseph. The context here is SSR for client component. Not SSG.

In SSR for client component, there's no way to pass security critical state or data such as authentication token. I want to pass data for SSR phase only, not propagate to client runtime for security reason.

**Reply by icyJoseph** [reply]
Right, all client components are SSR'd, we agree on that.

So if you have a Server component that renders a Client component child, and passes to it a piece of serialisable data, this gets bundled as part of the RSC data structure, and you want to avoid this, correct? 

Where I stop following you, is that the Server Component could derive from the sensitive data, say, a boolean, that signals whether or not to do something within the client component. Isn't that what you need to do?

**Reply by ellemedit** [reply]
So you mean we can opt out client component behavior? That could be an option but the approach is not good at reducing network roundtrip.

1. Server -> Client with Streaming Rendering
2. Server -> Client -> Fetch -> Server -> Client

So 2 can be a solution for a while, but I believe 1 is right here.

**Reply by icyJoseph** [reply]
I think we have different mental models of the "use case" at hand.

Plain and simple, there's an HttpOnly Cookie on your incoming request, and you want to do something, based off that cookie to a client component, correct? 

Here for example, https://github.com/icyJoseph/next-js-react-gbg/blob/feat/app-router-rsc/demo-app/src/app/collection/page.tsx, I calculate a Collection of Pokemon based off a HttpOnly Cookie, and then pass that collection to a client component.

Never does the client ever see the cookie interpretation (it uses a JWT decoding algorithm to calculate the collection). Is that not what you want? somewhat?

**Reply by ellemedit** [reply]
You right, Joseph.

Passing fetch promise client component from server component work.

Fetching data always should happen in server component, not client component itself.
I wanted to break their relation.

**Reply by ellemedit** [reply]
Thank you for noticing this again.
